### DISCUSSION TITLE
Setting connection for PSRP operator

### DISCUSSION THREAD

**Discussion by JaureguiOderda** [apache/airflow#27351] [Q&A] [answered]
Hello everyone,

I'm trying to set up a PsrpOperator to execute some powershell tasks on a Windows Server. Unfortunately in the documentation there is nothing about how to set up the connection to said server.

This is the code I have written, to execute a very simple command (Get-Item .) on the remote server:

```
import pendulum
from airflow import DAG
from airflow.operators.empty import EmptyOperator
from airflow.providers.microsoft.psrp.operators.psrp import PsrpOperator


with DAG(
    dag_id="Test_Powershell",
    description="Training pipeline",
    start_date=pendulum.datetime(2022, 10, 26, tz="Europe/Berlin"),
    catchup=False,
    tags=["Testing"],
) as dag:

    init = EmptyOperator(
        task_id="init",
    )


    new_pwsh = PsrpOperator(
        task_id="Testing_powershell_operator",
        psrp_conn_id="xxxxx",
        powershell="Get-Item .",
    )

    end = EmptyOperator(
        task_id="end",
    )

init >> new_pwsh >> end 

```

But the task fails, the log follows under here:

```
AIRFLOW_CTX_DAG_OWNER=airflow
AIRFLOW_CTX_DAG_ID=Test_Powershell
AIRFLOW_CTX_TASK_ID=Testing_powershell_operator
AIRFLOW_CTX_EXECUTION_DATE=2022-10-28T15:47:04.440881+00:00
AIRFLOW_CTX_TRY_NUMBER=1
AIRFLOW_CTX_DAG_RUN_ID=manual__2022-10-28T15:47:04.440881+00:00
[2022-10-28, 19:47:19 CEST] {base.py:68} INFO - Using connection ID 'xxxxx' for task execution.
[2022-10-28, 19:47:19 CEST] {psrp.py:123} INFO - Establishing WinRM connection xxxxx to host: xxxxx
[2022-10-28, 19:47:19 CEST] {powershell.py:133} INFO - Initialising RunspacePool object for configuration Microsoft.PowerShell
[2022-10-28, 19:47:19 CEST] {powershell.py:525} INFO - Opening a new Runspace Pool on remote host
[2022-10-28, 19:47:19 CEST] {local_task_job.py:156} INFO - Task exited with return code Negsignal.SIGSEGV
[2022-10-28, 19:47:19 CEST] {taskinstance.py:1415} INFO - Marking task as FAILED. dag_id=Test_Powershell, task_id=Testing_powershell_operator, execution_date=20221028T154704, start_date=20221028T154719, end_date=20221028T154719
[2022-10-28, 19:47:19 CEST] {local_task_job.py:273} INFO - 0 downstream tasks scheduled from follow-on schedule check

```

I'm quite sure that I'm not setting up the connection properly, but I have no connection like Windows Server, PSRP, Powershell, or anything similar. I've tried with HTTP and SSH, as well with generic and port 5689, but none of those worked.

Does anyone have an idea on how to set up the connection for this kind of tasks? What am I missing?

Thanks!

**Comment by potiuk** [comment]
(I know nothing about the operator - just looked at its code)

Sems the only thingi it uses is "host" so you can use any connection.

Look here: https://airflow.apache.org/docs/apache-airflow-providers-microsoft-psrp/stable/_api/airflow/providers/microsoft/psrp/operators/psrp/index.html

I believe you need to specify wsman_options - the help does not render it correctly but it is  `pypsrp.wsman.WSMan` that you should look for (external library). 

That's all that's visible in the code. I know @malthe worked on the operator quite a lot so maybe he can help.

**Reply by Taragolis** [reply]
Oh... I see in the provided log Segmentation Violation error.

You might also find some info in other discussion related to this (spoiler alert: this not related to Airflow, usual it refers to something wrong in the environment or libraries).
https://github.com/apache/airflow/discussions?discussions_q=Negsignal.SIGSEGV

**Reply by potiuk** [reply]
yeah Good catch (though I wonder @Taragolis  where dd you see the logs with segv ? I missed it :) 

Additionally - if you are on Apple M1 @DerDotkom - then it might be the reason - possibly the psrp library is not realy working well with it (we've seen a number of similar cases for other libraries)

**Reply by malthe** [reply]
If you can get a hold of a Python compiled with debug symbols then you can figure out where this segmentation comes from via gdb. But I don't think it's particularly related to `pypsrp`. Perhaps an XML library that it uses, those are sometimes using C-extensions.

**Reply by Taragolis** [reply]
@potiuk this one
`[2022-10-28, 19:47:19 CEST] {local_task_job.py:156} INFO - Task exited with return code Negsignal.SIGSEGV`

**Reply by potiuk** [reply]
:facepalm: INDEED :facepalm: ... Must have been blind :D

**Reply by Taragolis** [reply]
Might be also good point to add some huge warning if exit code `Negsignal.SIGSEGV` (or 11) with general information why it happen and how to debug it.

```
*************************************************************
Task exited with return code Negsignal.SIGSEGV
Usual if refers that something wrong with your environment
...
*************************************************************
```

**Reply by potiuk** [reply]
VERY GOOD IDEA!

**Reply by potiuk** [reply]
Go for it - happy to review it

**Comment by JaureguiOderda** [comment]
So additionally to running the airflow standalone in my M1 Mac, which might be a problem, I found an additional issue.

The server that I'm trying to connect to run this script is accessible from my laptop through RDP, but I cannot reach the port configured for powershell remoting (5985). From the other servers in the same subnet I can reach that port perfectly, but not from my laptop. There is some configuration there in the firewall from the network team preventing it for sure.

I will try pushing the code to the production airflow, which is located within the same network as our other servers far as I know, and try to discard network issues that way.

I will try this tests and I will write the outcome here.

Thank you all for your answers!!

**Reply by Taragolis** [reply]
@DerDotkom I've tried to run your sample DAG in Airflow in standalone mode:
 - macOS 13 (M1 Processor)
 - Python 3.9.9
 - Airflow 2.4.2 installed in fresh venv. 
 - `apache-airflow-providers-microsoft-psrp==2.0.0`
 - Use default configurations except Fernet Key and AIRFLOW_HOME
 - Additionally set  environment variable `PYTHONFAULTHANDLER` to `true` (`export PYTHONFAULTHANDLER=true`)
 
```shell
❯ airflow info

System info
OS              | Mac OS                                                                                                 
architecture    | arm                                                                                                    
uname           | uname_result(system='Darwin', node='REDACTED', release='22.1.0',    
                | version='Darwin Kernel Version 22.1.0: Sun Oct  9 20:15:09 PDT 2022;                                   
                | root:xnu-8792.41.9~2/RELEASE_ARM64_T6000', machine='arm64')                                            
locale          | (None, 'UTF-8')                                                                                        
python_version  | 3.9.9 (main, Jan  5 2022, 16:02:57)  [Clang 13.0.0 (clang-1300.0.27.3)]                                
python_location | /Users/taragolis/Projects/REDACTED/.nox/develop/bin/python  
```

And got the same error in task but in service logs I've got additional traceback (only when `PYTHONFAULTHANDLER` set to `true`)

```text
 scheduler | [2022-11-01 18:33:26,828] {scheduler_job.py:498} INFO - Setting the following tasks to queued state:
 scheduler | <TaskInstance: Test_Powershell.Testing_powershell_operator scheduled__2022-10-31T14:29:46.107570+00:00 [scheduled]>
 scheduler | [2022-11-01 18:33:26,829] {scheduler_job.py:537} INFO - Sending TaskInstanceKey(dag_id='Test_Powershell', task_id='Testing_powershell_operator', run_id='scheduled__2022-10-31T14:29:46.107570+00:00', try_number=3, map_index=-1) to executor with priority 1 and queue default
 scheduler | [2022-11-01 18:33:26,829] {base_executor.py:95} INFO - Adding to queue: ['airflow', 'tasks', 'run', 'Test_Powershell', 'Testing_powershell_operator', 'scheduled__2022-10-31T14:29:46.107570+00:00', '--local', '--subdir', 'DAGS_FOLDER/sample.dag.py']
 scheduler | [2022-11-01 18:33:26,830] {sequential_executor.py:61} INFO - Executing command: ['airflow', 'tasks', 'run', 'Test_Powershell', 'Testing_powershell_operator', 'scheduled__2022-10-31T14:29:46.107570+00:00', '--local', '--subdir', 'DAGS_FOLDER/sample.dag.py']
 scheduler | [2022-11-01 18:33:27,143] {dagbag.py:537} INFO - Filling up the DagBag from /Users/taragolis/Projects/REDACTED/.airflow_home/dags/sample.dag.py
 scheduler | [2022-11-01 18:33:27,401] {task_command.py:376} INFO - Running <TaskInstance: Test_Powershell.Testing_powershell_operator scheduled__2022-10-31T14:29:46.107570+00:00 [queued]> on host ip-10-192-244-47.eu-west-1.compute.internal
 scheduler | Fatal Python error: Segmentation fault
 scheduler | 
 scheduler | Current thread 0x00000001f1bb6500 (most recent call first):
 scheduler | File "/Users/taragolis/.pyenv/versions/3.9.9/lib/python3.9/urllib/request.py", line 2623 in proxy_bypass_macosx_sysconf
 scheduler | File "/Users/taragolis/.pyenv/versions/3.9.9/lib/python3.9/urllib/request.py", line 2647 in proxy_bypass
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/requests/utils.py", line 809 in should_bypass_proxies
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/requests/utils.py", line 825 in get_environ_proxies
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/requests/sessions.py", line 759 in merge_environment_settings
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/pypsrp/wsman.py", line 881 in _build_session
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/pypsrp/wsman.py", line 796 in send
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/pypsrp/wsman.py", line 470 in invoke
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/pypsrp/wsman.py", line 290 in create
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/pypsrp/shell.py", line 205 in open
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/pypsrp/powershell.py", line 548 in open
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/pypsrp/powershell.py", line 258 in __enter__
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/providers/microsoft/psrp/hooks/psrp.py", line 103 in __enter__
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/providers/microsoft/psrp/operators/psrp.py", line 125 in execute
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1664 in _execute_task
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1603 in _execute_task_with_callbacks
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1457 in _run_raw_task
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/utils/session.py", line 75 in wrapper
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/cli/commands/task_command.py", line 264 in _run_raw_task
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/cli/commands/task_command.py", line 191 in _run_task_by_selected_method
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/cli/commands/task_command.py", line 382 in task_run
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/utils/cli.py", line 103 in wrapper
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/cli/cli_parser.py", line 52 in command
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/task/task_runner/standard_task_runner.py", line 95 in _start_by_fork
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/task/task_runner/standard_task_runner.py", line 43 in start
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/jobs/local_task_job.py", line 107 in _execute
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/jobs/base_job.py", line 247 in run
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/cli/commands/task_command.py", line 247 in _run_task_by_local_task_job
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/cli/commands/task_command.py", line 189 in _run_task_by_selected_method
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/cli/commands/task_command.py", line 382 in task_run
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/utils/cli.py", line 103 in wrapper
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/cli/cli_parser.py", line 52 in command
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/lib/python3.9/site-packages/airflow/__main__.py", line 39 in main
 scheduler | File "/Users/taragolis/Projects/REDACTED/.nox/develop/bin/airflow", line 8 in <module>
 scheduler | [2022-11-01 18:33:27,612] {scheduler_job.py:589} INFO - Executor reports execution of Test_Powershell.Testing_powershell_operator run_id=scheduled__2022-10-31T14:29:46.107570+00:00 exited with status success for try_number 3
 scheduler | [2022-11-01 18:33:27,617] {scheduler_job.py:632} INFO - TaskInstance Finished: dag_id=Test_Powershell, task_id=Testing_powershell_operator, run_id=scheduled__2022-10-31T14:29:46.107570+00:00, map_index=-1, run_start_date=2022-11-01 14:33:27.426008+00:00, run_end_date=None, run_duration=6.653107, state=running, executor_state=success, try_number=3, max_tries=2, job_id=7, pool=default_pool, queue=default, priority_weight=2, operator=PsrpOperator, queued_dttm=2022-11-01 14:33:26.828740+00:00, queued_by_job_id=6, pid=10717
```

I realised that error happen in builtin `urllib.request` and found some solution (not sure is it permanent or not) in the other discussions and set environment variable `no_proxy` to `*` (`export no_proxy='*'`).
After that "Segmentation fault" gone.

Also I tried to find something in CPython issues:
- https://github.com/python/cpython/issues/58037 (most of the same issues refers to this one)
- https://github.com/python/cpython/issues/71313
- https://wefearchange.org/2018/11/forkmacos.rst.html

cc: @potiuk

**Reply by potiuk** [reply]
I think there were other cases when NO_PROXY solved the problem.  So I guess that's it.

**Reply by Taragolis** [reply]
Might be better in case of run in macOS try to not use fork in task runner or make it configurable by something like `[core] macos_task_fork` = `True | False`. But it just an idea which might also has some side effects.

**Reply by potiuk** [reply]
I think there are too many side effects to worry. We already have https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#execute-tasks-new-python-interpreter but turning it on by default on Mac means that you are slowing down everything A LOT and you give impression to the MacOS users that Airflow is far slower in dependency processing than it is. I think there are a number of other cases which will cause Mac to produce SIGSEGV. I would rather instead add a note about this "option" to the PR of yours https://github.com/apache/airflow/pull/27381  (which BTW would be great to finalize :) )

**Reply by Taragolis** [reply]
I've actually also tried to set this config to `True` but with no luck and got the same error. I thought it related to the fact that this configuration affect Executors but `StandardTaskRunner` might fork process anyway

https://github.com/apache/airflow/blob/6434b5770877d75fba3c0c49fd808d6413367ab4/airflow/task/task_runner/standard_task_runner.py#L41-L45

**Reply by potiuk** [reply]
Ah. I see. It amuses me sometimes how many times we fork :) . Probably setting runner to CGroupTaskRunner might help (as a workaround/test at least whether it will work)

It almost calls for a new config parameter, but I would like to avoid it. However why don't we do it differently. Why don't we add another task runner that will be extendingthe StandardTaskRunner but it will never fork? NonForkingTaskRunner (?) - this might be a nice and simple solution without having to add a new configuration parametter and we will be able to avoid the problem ? 

This will be the body of it:

```
    def start(self):
        self.process = self._start_by_exec()
```

**Comment by malthe** [comment]
@DerDotkom note that if you want to use 5985, then you really must use Kerberos as the authentication protocol, otherwise it is unsafe.

**Reply by JaureguiOderda** [reply]
Good point. Is that any configuration that I need to make from the code, while configuring the operator? Or is is a server/network-level configuration?

**Reply by malthe** [reply]
The default is 5986 because the default value for `ssl` in `pypsrp.wsman.WSMan` is true.

But you can override this value using `wsman_options`.

**Comment by JaureguiOderda** [comment, marked as answer]
I was able to make it work! Thank you so much for your help!
The code ended up looking like this. Now I have to speak to the sysadmins so they can set up kerberos for the authentication as weak for the second hop in powershell, and avoid the cached credentials from CredSSP. But at least I already have enough to leave some pipelines working while we tackle this kerberos deployment.

```
import pendulum
from airflow import DAG
from airflow.operators.empty import EmptyOperator
from airflow.providers.microsoft.psrp.operators.psrp import PsrpOperator


with DAG(
    dag_id="Test_Powershell",
    description="Training pipeline",
    start_date=pendulum.datetime(2022, 10, 26, tz="Europe/Berlin"),
    catchup=False,
    tags=["Testing"],
) as dag:

    init = EmptyOperator(
        task_id="init",
    )

    new_pwsh = PsrpOperator(
        task_id="Testing_powershell_operator",
        psrp_conn_id="XXXXXX", 
        powershell='Write-Output "Hello!" | Out-File "\\\\some\\network\\file.txt" ',
        wsman_options={"ssl": False, "auth": "credssp"},
    )

    end = EmptyOperator(
        task_id="end",
    )

init >> new_pwsh >> end

```

I want to thank you all again, getting it to work without your help would have been so much more complicated. :)

**Comment by malthe** [comment]
@DerDotkom just remembering now that I did a write-up about this operator: https://www.maltheborch.com/2022/02/powershell-remoting-on-windows-using-airflow.html.
